### ISSUE TITLE
To cache or not to cache…

### ISSUE THREAD

**Issue by levifig** [levifig/now-playing.widget#2]
Ideally, we don't make requests to Spotify's API every 5 seconds. The main reason why I'm calling the API is to get the album artwork. I can get the track information (e.g. album, artist, track name) from the app directly (using Applescript) but can't AFAIK get album artwork.

While I would like to cache the album artwork (using the album ID, e.g. `cache/asdhASD347yasd.jpg`) and have the CSS pull the image from there, that would bring up the issue of cache size, updates, etc.

The simplest way to reduce requests to the API would be to poll the app every X seconds but only poll the API **if** the track changed. That would significantly reduce API requests and avoid dealing with local cache. The downside to this would be the lack of offline support, something that a local cache would allow. While the images are fairly small (the 640px version is ~75kb, and the 300px on is ~25kb, though it would look bad on a @2x display since I'm displaying it as 200px), it would still use a substancial amount of storage space after some time.

Decisions, decisions…
Feedback **definitely** welcome and encouraged! :)

**Comment by ntucakovic** [comment]
Very nice to see someone working on a proper track info. I haven't worked in coffeescript that much and not sure of possibilities. I feel that the idea to keep track id and only check if it's changed is completely fine. We could also set up a config option whether to use local caching or not (at some point). 
I'll have a look these days when I find a bit of time because I feel like this caching is definitely a priority :)

**Comment by levifig** [comment]
@ntucakovic Thank you, sir! :D I'm super swamped ATM, but I plan on revisiting this widget very soon, especially considering some annoying bugs have surfaced since the last Übersicht update… :\

I look forward to brainstorming caching strategies in the future… ;)

**Comment by evanfuture** [comment]
Ok, just a quick thought.  When you download the artwork, could you name it the same as the track id, and then just check if there's already a file in the cache matching before you re-download it?

**Comment by levifig** [comment]
@evanfuture That's definitely what I had in mind. The only problem is that we would then have to manage the cache: how big should we let it get? how old? etc… This is especially noticeable for people that listen to a lot of playlists (i.e. lots of different album artwork).

If we find a way to manage the cache, that strategy works well. :) Thoughts?

**Comment by ntucakovic** [comment]
I haven't looked into responses from spotify API, but seems like that ain't the ideal solution? Do the tracks from the same album have same urls to album arts? I think it would be better to cache it with album art url (or some other identifier). I'll definitely have a go at it this weekend, if someone else doesn't do it sooner :)

**Comment by evanfuture** [comment]
Seems like the goal is to mainly to reduce the amount of calls, so even clearing the cache once per week or day wouldn't lead to that large a cache, or that much re-downloading.

Also, any luck getting Ubersicht to stop returning Error messages all the time?

**Comment by levifig** [comment]
@evanfuture Good point. A simple implementation like that could be a great first step.

As for the errors, I'm hoping felixhageloh/uebersicht#171 solves that… soon… :) but if not, I'll have to figure something else out! As it stands, I'm pretty sure the error is caused by Übersicht trying to run a script I include as a standalone script, which obviously won't work! :P

**Comment by levifig** [comment]
Has anyone seen any rate limiting on this? I've been using this widget for months now and I have never hit any sort of rate limiting… Since that simplifies this widget (caching would be one giant undertaking and inevitably increase bugs and reduce maintainability), I'm gonna keep it uncached for now. I'll reconsider this eventually, should Spotify change its API policies.

Feel free to report any issues you may be experiencing. 👍 
Thanks!
